### ISSUE TITLE
`@OptimisticLocking(DIRTY)` not working with `@OneToOne`

### ISSUE THREAD

**Issue by mickroll** [quarkusio/quarkus#34155]
### Describe the bug

Having two entities:
```
@DynamicUpdate
@OptimisticLocking(DIRTY)
public class Basket {
    [...]
    @OneToOne(cascade = ALL, fetch = LAZY, orphanRemoval = true)
    private Fruit fruit;
}

@DynamicUpdate
@OptimisticLocking(DIRTY)
public class Fruit {
    [...]
}
```
... and updating the fruit through the basket, leads to incomplete parameters for the dynamic update.

### Expected behavior

dirty update also works for referenced entities through cascading

### Actual behavior

```
org.hibernate.exception.DataException: could not execute statement [Für den Parameter 4 wurde kein Wert angegeben.] [update Fruit set name=?,type=? where id=? and name=? and type=?]
[...]
Caused by: org.postgresql.util.PSQLException: Für den Parameter 4 wurde kein Wert angegeben.
	at org.postgresql.core.v3.SimpleParameterList.checkAllParametersSet(SimpleParameterList.java:284)
```
The german message means "no value defined for parameter 4", which is the first additional parameter that dirty optimistic locking adds to the where clause.

### How to Reproduce?

https://github.com/mickroll/quarkus-demonstrator/tree/optimistic-locking-dirty

Steps:
- run `mvn test`
- BasketTest fails
- check stacktraces

### Output of `uname -a` or `ver`

ubuntu x86_64

### Output of `java -version`

openjdk version "17.0.7" 2023-04-18 LTS

### Quarkus version or git rev

3.1.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

maven 3.8.6

**Comment by famod** [comment]
Btw, this has been working fine in Quarkus 2.16/Hibernate 5.6!

**Comment by geoand** [comment]
This is probably an ORM issue (and not a Quarkus specific one), but let's see what the SMEs say

**Comment by famod** [comment]
> This is probably an ORM issue (and not a Quarkus specific one), but let's see what the SMEs say

Yeah, probably.
We have more coming. @geoand do you want us create issues directly in the Hibernate JIRA instead?
It's much easier for us to create Quarkus based reproducers though and sometimes the Quarkus integration does play a role.

**Comment by geoand** [comment]
I think starting with issues here is fine and makes it easier for our SMEs to drill down to where the problem is

**Comment by yrodiere** [comment]
Our SMEs would particularly like if bugs in Hibernate ORM were reported to Hibernate ORM directly.

In an ideal world you'd try to reproduce the problem with Hibernate ORM only (see https://github.com/hibernate/hibernate-test-case-templates/tree/main/orm/hibernate-orm-6/) and if you can't, you'd report the problem here so that we'd investigate what's so specific about Quarkus exactly.

Thanks for the report, we'll have a look and will probably forward to the ORM team.

**Comment by famod** [comment]
@yrodiere do you have any news for us? Thanks!

**Comment by yrodiere** [comment]
No, I was busy with a deadline and couldn't look into this. @marko-bekhta now that we've (almost) released Search maybe you have some time to investigate?

**Comment by marko-bekhta** [comment]
Thanks for reporting the problem. It turns out to be an issue in ORM. Reported upstream - https://hibernate.atlassian.net/browse/HHH-16839. 
If you were trying to reproduce it without Quarkus and weren't able to -- it might be that you haven't enabled bytecode enhancement. A combination of lazy one-to-one association and enabled bytecode enhancement causes that error.

**Comment by famod** [comment]
Thank you both! Looking forward to a fix to continue our Quarkus 3 migration efforts.

**Comment by mickroll** [comment]
I would not have thought about turning on bytecode enhancement. Thanks for the upstream report!

**Comment by famod** [comment]
FWIW:

> Our SMEs would particularly like if bugs in Hibernate ORM were reported to Hibernate ORM directly :)
> 
> In an ideal world you'd try to reproduce the problem with Hibernate ORM only (see https://github.com/hibernate/hibernate-test-case-templates/tree/main/orm/hibernate-orm-6/) and if you can't, you'd report the problem here so that we'd investigate what's so specific about Quarkus exactly.

We filed our next issue (which blocks our Q3 upgrade) directly: https://hibernate.atlassian.net/browse/HHH-16845
...and yes, I'm trying to make some noise here. :wink:

**Comment by mickroll** [comment]
Upstream Hibernate fix targets version 6.2.6, see https://hibernate.atlassian.net/browse/HHH-16839

**Comment by geoand** [comment]
@yrodiere can this now be closed?
